File: mailblastr-python/mailblastr/logs.py
"""API request logs resource."""

from typing import TypedDict

from . import http_client
from .http_client import build_query, path_escape as _e


class ListParams(TypedDict, total=False):
    limit: int
    after: str
    before: str
    method: str  # filter to an exact HTTP method, e.g. 'POST'
    status: int  # filter to an exact response status, e.g. 200 or 429


class Logs:
    """``mailblastr.Logs`` — the /logs endpoints (read-only)."""

    ListParams = ListParams

    @classmethod
    def list(cls, params=None):
        """List API request logs. Cursor-paginated (``limit``/``after``/
        ``before``) with optional ``method`` and ``status`` filters. GET /logs"""
        params = params or {}
        qs = build_query(
            {
                "limit": params.get("limit"),
                "after": params.get("after"),
                "before": params.get("before"),
                "method": params.get("method"),
                "status": params.get("status"),
            }
        )
        return http_client.request("GET", f"/logs{qs}")

    @classmethod
    def get(cls, log_id):
        """Retrieve a log entry (with request/response bodies). GET /logs/:id"""
        return http_client.request("GET", f"/logs/{_e(log_id)}")

File: mailblastr-rust/src/services/logs.rs
//! `mailblastr.logs` — read-only API request logs.

use std::sync::Arc;

use reqwest::Method;
use serde::Deserialize;
use serde_json::Value;

use crate::client::{seg, Config};
use crate::types::{ListResponse, Result};

/// One API request log entry.
#[derive(Debug, Clone, Deserialize)]
pub struct LogEntry {
    pub object: String,
    pub id: String,
    pub endpoint: Option<String>,
    pub method: String,
    pub response_status: u16,
    pub user_agent: Option<String>,
    pub created_at: String,
    /// Present only on retrieve (`GET /logs/:id`).
    pub request_body: Option<Value>,
    /// Present only on retrieve (`GET /logs/:id`).
    pub response_body: Option<Value>,
}

/// Params for `logs.list`: cursor pagination plus optional server-side
/// `method`/`status` filters.
#[derive(Debug, Clone, Default)]
pub struct ListLogsParams {
    pub limit: Option<u32>,
    pub after: Option<String>,
    pub before: Option<String>,
    /// Filter to an exact HTTP method, e.g. `POST`.
    pub method: Option<String>,
    /// Filter to an exact response status, e.g. `200` or `429`.
    pub status: Option<u16>,
}

impl ListLogsParams {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_limit(mut self, limit: u32) -> Self {
        self.limit = Some(limit);
        self
    }

    pub fn with_after(mut self, after: impl Into<String>) -> Self {
        self.after = Some(after.into());
        self
    }

    pub fn with_before(mut self, before: impl Into<String>) -> Self {
        self.before = Some(before.into());
        self
    }

    pub fn with_method(mut self, method: impl Into<String>) -> Self {
        self.method = Some(method.into());
        self
    }

    pub fn with_status(mut self, status: u16) -> Self {
        self.status = Some(status);
        self
    }
}

/// `mailblastr.logs`.
#[derive(Clone, Debug)]
pub struct LogsSvc {
    config: Arc<Config>,
}

impl LogsSvc {
    pub(crate) fn new(config: Arc<Config>) -> Self {
        Self { config }
    }

    /// List API request logs (cursor-paginated, optional filters).
    /// `GET /logs`
    pub async fn list(&self, params: Option<ListLogsParams>) -> Result<ListResponse<LogEntry>> {
        let mut query: Vec<(&str, String)> = Vec::new();
        if let Some(params) = &params {
            if let Some(limit) = params.limit {
                query.push(("limit", limit.to_string()));
            }
            if let Some(after) = &params.after {
                query.push(("after", after.clone()));
            }
            if let Some(before) = &params.before {
                query.push(("before", before.clone()));
            }
            if let Some(method) = &params.method {
                query.push(("method", method.clone()));
            }
            if let Some(status) = params.status {
                query.push(("status", status.to_string()));
            }
        }
        self.config
            .send(self.config.request(Method::GET, "/logs").query(&query))
            .await
    }

    /// Retrieve a log entry (includes request/response bodies).
    /// `GET /logs/:id`
    pub async fn get(&self, log_id: &str) -> Result<LogEntry> {
        let path = format!("/logs/{}", seg(log_id));
        self.config
            .send(self.config.request(Method::GET, &path))
            .await
    }
}
